machinelearningteam/summary-statistics-service/README.md

```diff
@@ -0,0 +1,83 @@
+...
+```
+
+To keep up with different document formats, we want the service to automatically exclude columns that have too many different values from the summary table and we want to be able to specify other columns to aggregate by. 
+
+
+## The sample code
+We have provided the code for a simple gRPC API for the statistics service, along with YAML manifests for deploying the API to a Kubernetes cluster. The code currently just echos the expenses sent in the request, no statistics done yet.
+
+### Prerequisites
+You need to have a working setup for writing code in Go, building container images with Docker, and deploying to a Kubernetes cluster with eg. `kubectl`
+
+### Local testing
+The sample is written in Go, with the API defined by the Protocol Buffer found in `proto`. You can start the API locally by running:
+
+```go run cmd/service/service.go```
+
+And test it out with the client:
+
+```go run cmd/client/client.go```
+
+This will take the document in `test.csv`, send it to the service and and save the output to `out.csv`
+
+### Deploy on Kubernetes
+Firstly you need to build the Docker image and push it to some image registry that your cluster has access to. To deploy the service on a Kubernetes cluster use the manifests in `manifests`. The service and deployment in `api.yaml` should spin up the API service and get you an external ip to connect to.
+
+## Your Task
+We would like you to extend this system with actual statistics, and make the system more "production ready".
+
+### Part 1 - The mandatory part
+Create you own micro-service to do the statistics processing, separate from the API service. You can choose to write it in whatever programming language you are comfortable with. We expect it to:
+
+* Contain an API definition in a separate protobuf file
+* Handle requests through gRPC
+* Take as part of the request:
+    * The document originally sent to the API
+    * Parameters for aggregation columns
+    * Parameters for computing statistics over columns and columns to exclude
+
+Alongside your code you should also write a manifest YAML to setup a deployment and service. The deployment should specify that your container should be replicated at least in 2 pods, which should be load-balanced through the service.
+
+You will also have to add code in the existing API code in `pkg/api/api.go`, so that it will forward the document to your service. You should not have to change the existing API and protobuf. The API should satisfy the following:
+
+* Be configurable, such that the desired aggregation options should be supplied at startup. How is up to you. You can use 25 values as the cut-off for categorical columns.
+* Have an option to supply the API with a URL for downloading documents, instead of documents bytes(This is already present in the protobuf, but not implemented). Whether you want to download the document content in the API and send bytes to your service - or send the URL and download in the service - is up to you.
```

Comment:
Leftover from the old assignment?

Reply:
The URL part? Yeah, it doesn't translate super well unless we interpret URLs to include storage buckets.

Reply:
It still forces the decision of where to initiate the download.